tests/worker/test_workflow.py

```diff
@@ -6439,3 +6439,83 @@ async def test_concurrent_sleeps_use_proper_options(
 
         # Force replay with a query to ensure determinism
         await handle.query("__temporal_workflow_metadata")
+
+
+@workflow.defn
+class SignalsActivitiesTimersUpdatesTracingWorkflow:
+    def __init__(self) -> None:
+        self.events: List[str] = []
+
+    @workflow.run
+    async def run(self) -> List[str]:
+        tt = asyncio.create_task(self.run_timer())
+        at = asyncio.create_task(self.run_act())
+        await asyncio.gather(tt, at)
+        return self.events
+
+    @workflow.signal
+    async def dosig(self, name: str):
+        self.events.append(f"sig-{name}-sync")
+        fut: asyncio.Future[bool] = asyncio.Future()
+        fut.set_result(True)
+        await fut
+        self.events.append(f"sig-{name}-1")
+        await workflow.wait_condition(lambda: True)
+        self.events.append(f"sig-{name}-2")
+
+    @workflow.update
+    async def doupdate(self, name: str):
+        self.events.append(f"update-{name}-sync")
+        fut: asyncio.Future[bool] = asyncio.Future()
+        fut.set_result(True)
+        await fut
+        self.events.append(f"update-{name}-1")
+        await workflow.wait_condition(lambda: True)
+        self.events.append(f"update-{name}-2")
+
+    async def run_timer(self):
+        self.events.append("timer-sync")
+        await workflow.sleep(0.1)
+        fut: asyncio.Future[bool] = asyncio.Future()
+        fut.set_result(True)
+        await fut
+        self.events.append("timer-1")
+        await workflow.wait_condition(lambda: True)
+        self.events.append("timer-2")
+
+    async def run_act(self):
+        self.events.append("act-sync")
+        await workflow.execute_activity(
+            say_hello, "Enchi", schedule_to_close_timeout=timedelta(seconds=5)
+        )
+        fut: asyncio.Future[bool] = asyncio.Future()
+        fut.set_result(True)
+        await fut
+        self.events.append("act-1")
+        await workflow.wait_condition(lambda: True)
+        self.events.append("act-2")
+
+
+async def test_async_loop_ordering(client: Client):
```

Comment:
Hrmm, I think another test may look like this:

* Create workflow that has a signal handler that updates some counter and an activity in the background that, when done, updates that same counter. And have the primary workflow method return the counter when it is `> 2` (i.e. after wait condition for that).
* Start the workflow, and while the activity is running, kill the workflow worker
* While the workflow worker is not running, send two signals and complete the activity
* Start the workflow worker again and confirm result.

I think with older code, the counter will only reach `2` before workflow returns (i.e. the two signals), but this newer code may return `3` (because it completed the activity _before_ it processed wait condition). I have not tested this though.

Reply:
(I assume you meant >=2, otherwise it's always the same). Bleh, yes indeed that does cause a difference, makes sense.

Well, then we've gotta decide if this is worth a flag or not too. Somewhat more worth it maybe since the batches are kind of odd, but, if we're going to have to keep that code around forever anyway... :man_shrugging:. What do you think?

At minimum these tests are kind of useful to keep around.

Reply:
To be honest, I think the benefit of code/logic cleanup goes away if you have to keep the old stuff around. But I agree the test is very valuable even as a separate PR.

Reply:
This is where it's relevant that we have other changes that we'd like to make that will also require a flag I think?

>  I think the benefit of code/logic cleanup goes away if you have to keep the old stuff around.

I don't think I agree with that. For readers of the code it's an improvement to be given the "nicest" implementation to read, and also be referred to the location of a previous implementation that can be hidden away somewhat.

Reply:
Readers of code will have two implementations to read which is worse, and I still think it's not worth it to help readers of code if there is any runtime penalty (such as maintaining two forked pieces of logic). Having said that, I agree if we solved the wait condition issue at the same time this could be worth it.

Reply:
> Readers of code will have two implementations to read which is worse

I'm not sure what you mean. I just gave my opinion that having the old impl around was a net improvement, and obviously when I formed that opinion I took into account that there would be two implementations around!

Reply:
Also, if you read what I wrote you'll see that my point is that readers don't need to read the old implementation. They can read the new one, and unless they have reason to be interested in it, take it on trust that the old one is equivalent.

Reply:
What I'm saying is that I don't think the cost of having two implementations around is worth the benefit of simplifying one of them. I also don't think we should have runtime side effects on users, e.g. SDK flags in history, if the only benefit is internal code legibility. But maybe we can combine with actual user benefit.

Reply:
> What I'm saying is that I don't think the cost of having two implementations around is worth the benefit of simplifying one of them.

Yes, you already gave that opinion, and then I gave a countering opinion. After that point neither of us have added anything to the conversation; in that situation there's no need for us to compose further replies.

>  I also don't think we should have runtime side effects on users, e.g. SDK flags in history, if the only benefit is internal code legibility. But maybe we can combine with actual user benefit.

Yes, I already said that:

> This is where it's relevant that we have other changes that we'd like to make that will also require a flag I think?

Reply:
I agree with Chad I think. Having two is worse than one nicer and one old, I think.

If you want to look at the "nice" implementation, you can look at Ruby now as a reference so that's somewhat redeeming.

I can split out the tests to another PR.

Reply:
Sounds good. I'm definitely going to keep in mind that we could bundle this with improving `wait_condition`.